contracts/staking/PoSValidatorManager.sol

```diff
@@ -337,52 +337,57 @@ abstract contract PoSValidatorManager is IPoSValidatorManager, ValidatorManager
 
         // Ensure the delegator is active
         Delegator memory delegator = $._delegatorStakes[delegationID];
+        Validator memory validator = getValidator(validationID);
         require(
             delegator.status == DelegatorStatus.Active, "PoSValidatorManager: delegation not active"
         );
         // Only the delegation owner can end the delegation.
         require(
             delegator.owner == _msgSender(), "PoSValidatorManager: delegation not owned by sender"
         );
-        uint64 nonce = _incrementAndGetNonce(validationID);
 
         // Set the delegator status to pending removed, so that it can be properly removed in
         // the complete step, even if the delivered nonce is greater than the nonce used to
         // initialize the removal.
         delegator.status = DelegatorStatus.PendingRemoved;
-        delegator.endedAt = uint64(block.timestamp);
-        delegator.endingNonce = nonce;
-
-        $._delegatorStakes[delegationID] = delegator;
-
-        Validator memory validator = getValidator(validationID);
-        require(validator.weight > delegator.weight, "PoSValidatorManager: Invalid weight");
 
-        // Check that removing this delegator would not exceed the maximum churn rate.
-        // We only need to check this is the validator is still active. If the validator ends its validation
-        // period, the weight of all its delegators will be added to the churn tracker at that time. Ending
-        // a delegation whose validator has ended validating has no impact on the stake weight of the chain.
         if (validator.status == ValidatorStatus.Active) {
+            // Check that removing this delegator would not exceed the maximum churn rate.
+            // We only need to check this is the validator is still active. If the validator ends its validation
+            // period, the weight of all its delegators will be added to the churn tracker at that time. Ending
+            // a delegation whose validator has ended validating has no impact on the stake weight of the chain.
             _checkAndUpdateChurnTrackerRemoval(delegator.weight);
-        }
 
-        uint64 newValidatorWeight = validator.weight - delegator.weight;
-        _setValidatorWeight(validationID, newValidatorWeight);
+            delegator.endingNonce = _incrementAndGetNonce(validationID);
+            delegator.endedAt = uint64(block.timestamp);
+
+            uint64 newValidatorWeight = validator.weight - delegator.weight;
+            _setValidatorWeight(validationID, newValidatorWeight);
+
+            // Submit the message to the Warp precompile.
+            bytes32 messageID = WARP_MESSENGER.sendWarpMessage(
+                ValidatorMessages.packSetSubnetValidatorWeightMessage(
+                    validationID, delegator.endingNonce, newValidatorWeight
+                )
+            );
+
+            emit ValidatorWeightUpdate({
```

Comment:
Would it make sense to move this emission to inside the `_setValidatorWeight` implementation? It may require removing the nonce/messageID fields, but feels a bit less error prone (as is, I think we're forgetting to emit it in the `initializeDelegatorRegistration` case).

Reply:
I've moved move of the weigh change logic to `_setValidatorWeight`, including checking the churn tracker. The only other time the churn tracker needs to be called is on validator registration.